File: CPP/lcm.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>

using namespace std;

int gcdIterativeOptimized(int a, int b){
    while (b){
        int tmp = a;
        a = b;
        b = tmp % b;
    }
    return a;
}

int gcdRecursiveOptimized(int a, int b){
    if (!b){
        return a;
    }
    return gcdRecursiveOptimized(b, a % b);
}

int lcm_using_gcd(int a, int b){
    // return a * b / gcdIterativeOptimized(a, b);
    return a * b / gcdRecursiveOptimized(a, b);
}

vector<int> factorizeEfficientVector(int number){
    vector<int> factors;
    while (number % 2 == 0){
        factors.push_back(2);
        number /= 2;
    }
    for (int factor = 3; factor <= sqrt(number); factor += 2){
        while (number % factor == 0){
            factors.push_back(factor);
            number /= factor;
        }
    }
    if (number > 1){
        factors.push_back(number);
    }
    return factors;
}

int lcmUsingPrimeFactors(int a, int b){
    vector<int> factors_a = factorizeEfficientVector(a);
    vector<int> factors_b = factorizeEfficientVector(b);
    int count_a, count_b, max_count;
    int result = 1;
    int factor = 2;
    while(result <= max(a, b)){
        count_a = count(factors_a.begin(), factors_a.end(), factor);
        count_b = count(factors_b.begin(), factors_b.end(), factor);
        max_count = max(count_a, count_b);
        result *= int(pow(factor, max_count));
        if(a % result == 0 && b % result == 0 && result != 1){
            return result;
        }
        factor++;
    }
    return result;
}

int main(int argc, char** argv) {
    cout << lcm_using_gcd(12, 18) << endl;
    cout << lcm_using_gcd(32, 28) << endl;

    cout << lcmUsingPrimeFactors(12, 18) << endl;
    cout << lcmUsingPrimeFactors(32, 28) << endl;
}

File: Python/lcm.py
#O(logn)
def gcd_iterative_optimized(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return a

#O(logn)
def gcd_recursive_optimized(a: int, b: int) -> int:
    if b:
        return gcd_recursive_optimized(b, a % b)
    return a

#O(logn)
def lcm_using_gcd(a: int, b: int) -> int:
    # return a * b // gcd_iterative_optimized(a, b)
    return a * b // gcd_recursive_optimized(a, b) 

print(lcm_using_gcd(12, 18))
print(lcm_using_gcd(28, 32))

#O(sqrt(n))
def factorize_efficient_list(number: int) -> list:
    factors = []
    while number % 2 == 0:
        factors.append(2)
        number //= 2
    factor = 3
    while factor * factor <= number:
        while number % factor == 0:
            factors.append(factor)
            number //= factor
        factor += 2
    if number > 1:
        factors.append(number)
    return factors

#O(sqrt(n))
def lcm_using_prime_factors(a: int, b: int) -> int:
    factors_a = factorize_efficient_list(a)
    factors_b = factorize_efficient_list(b)
    result = 1
    factor = 2
    while result <= max(a, b):
        count_a = factors_a.count(factor)
        count_b = factors_b.count(factor)
        max_count = max(count_a, count_b)
        result *= int(factor ** max_count)
        if a % result == 0 and b % result == 0 and result != 1:
            return result
        factor += 1
    return result

print(lcm_using_prime_factors(12, 18))
print(lcm_using_prime_factors(28, 32))